drivers/sensor/tmag5273/tmag5273.c

```diff
@@ -0,0 +1,1121 @@
+					      : TMAG5273_TRIGGER_MODE_I2C;
+	regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_TRIGGER_MODE_MSK, val);
+
+	val = (drv_cfg->operation_mode == TMAG5273_DT_OPER_MODE_CONTINUOUS)
+		      ? TMAG5273_OPERATING_MODE_CONTINUOUS
+		      : TMAG5273_OPERATING_MODE_STANDBY;
+	regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_OPERATING_MODE_MSK, val);
+
+	retval = i2c_reg_write_byte_dt(&drv_cfg->i2c, TMAG5273_REG_DEVICE_CONFIG_2, regdata);
+
+	if (retval < 0) {
+		LOG_ERR("error setting DEVICE_CONFIG_2");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief sets the \c SENSOR_CONFIG_1 and \c SENSOR_CONFIG_2 registers
+ *
+ * @param drv_cfg configuration of the TMAG5273 instance
+ *
+ * @retval 0 if everything was okay
+ * @retval -EIO on communication errors
+ */
+static inline int tmag5273_init_sensor_settings(const struct tmag5273_config *drv_cfg)
+{
+	int retval;
+
+	/* REG_SENSOR_CONFIG_1 */
+	uint8_t regdata = FIELD_PREP(TMAG5273_MAG_CH_EN_MSK, TMAG5273_MAGNETIC_CH_NONE);
+	uint8_t val = ((drv_cfg->x_axis) ? TMAG5273_MAGNETIC_CH_X_EN : TMAG5273_MAGNETIC_CH_NONE) |
+		      ((drv_cfg->y_axis) ? TMAG5273_MAGNETIC_CH_Y_EN : TMAG5273_MAGNETIC_CH_NONE) |
+		      ((drv_cfg->z_axis) ? TMAG5273_MAGNETIC_CH_Z_EN : TMAG5273_MAGNETIC_CH_NONE);
+
+	regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_MAG_CH_EN_MSK, val);
+
+	retval = i2c_reg_write_byte_dt(&drv_cfg->i2c, TMAG5273_REG_SENSOR_CONFIG_1, regdata);
+	if (retval < 0) {
+		LOG_ERR("error setting SENSOR_CONFIG_1");
+		return -EIO;
+	}
+
+	/* REG_SENSOR_CONFIG_2 */
+	regdata = 0;
+
+	val = (drv_cfg->ch_mag_gain_correction == TMAG5273_DT_CORRECTION_CH_2)
+		      ? TMAG5273_MAG_GAIN_CORRECTION_CH_2
+		      : TMAG5273_MAG_GAIN_CORRECTION_CH_1;
+
+	regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_GAIN_CORRECTION_CH_MSK, val);
+
+	switch (drv_cfg->angle_magnitude_axis) {
+	case TMAG5273_DT_ANGLE_MAG_XY:
+		val = TMAG5273_ANGLE_EN_XY;
+		break;
+	case TMAG5273_DT_ANGLE_MAG_YZ:
+		val = TMAG5273_ANGLE_EN_YZ;
+		break;
+	case TMAG5273_DT_ANGLE_MAG_XZ:
+		val = TMAG5273_ANGLE_EN_XZ;
+		break;
+	case TMAG5273_DT_ANGLE_MAG_RUNTIME:
+	case TMAG5273_DT_ANGLE_MAG_NONE:
+		__fallthrough;
+	default:
+		val = TMAG5273_ANGLE_EN_NONE;
+		break;
+	}
+
+	regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_MAG_CH_EN_MSK, val);
+
+	/* measurement range "high" by default */
+	switch (drv_cfg->meas_range) {
+	case TMAG5273_DT_AXIS_RANGE_LOW:
+		regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_X_Y_RANGE_MSK,
+					     TMAG5273_XYZ_MEAS_RANGE_LOW);
+		regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_Z_RANGE_MSK,
+					     TMAG5273_XYZ_MEAS_RANGE_LOW);
+		break;
+	case TMAG5273_DT_AXIS_RANGE_HIGH:
+	case TMAG5273_DT_AXIS_RANGE_RUNTIME:
+		__fallthrough;
+	default:
+		regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_X_Y_RANGE_MSK,
+					     TMAG5273_XYZ_MEAS_RANGE_HIGH);
+		regdata = TMAG5273_FIELD_SET(regdata, TMAG5273_Z_RANGE_MSK,
+					     TMAG5273_XYZ_MEAS_RANGE_HIGH);
```

Comment:
It'd be nice if you could avoid using custom access macros for single bit registers, may simplify the header a bit as well if you drop some MSK/POS pairs, something like this:

```c
#define TMAG5273_XYZ_MEAS_RANGE_LOW_MASK (BIT(TMAG5273_X_Y_RANGE) | BIT(TMAG5273_Z_RANGE))
...
regdata |= TMAG5273_XYZ_MEAS_RANGE_LOW_MASK;
```

Reply:
I would prefer to leave the header "as is", representing the registers as defined by the datasheet. But you are right,
- dropped the special setter macro nearly everywhere (was actually a fragment),
- specialed the range define on top of the *.c file.